### PR TITLE
feat(gensupport): per-chunk transfer timeout configs

### PR DESCRIPTION
**Pull request by Tulsishah**
feat(gensupport): per-chunk transfer timeout configs

Allow users to configure the per-chunk transfer timeout for retries
that's used during resumable uploads.

Needs to be exposed via the manual layer for storage.

Tested the feature(with default timeout and with some random value) with storagetestbench server emulator.

### CODE REVIEW

File: internal/gensupport/retry.go
Comment (tritone) [inline internal/gensupport/retry.go:30]: Default should be no timeout.

File: internal/gensupport/resumable_test.go
Comment (tritone) [inline internal/gensupport/resumable_test.go:283]: Why is this change needed? Can you use errors.Is?

File: internal/gensupport/resumable.go
Comment (tritone) [inline internal/gensupport/resumable.go:266]: How do we distinguish between a timeout in ctx vs rCtx?

I think we'd only want to retry if we get deadline exceeded from rCtx, but ctx is still good.

File: internal/gensupport/resumable_test.go
Comment (Tulsishah) [reply internal/gensupport/resumable_test.go]: Done.

File: internal/gensupport/resumable.go
Comment (Tulsishah) [reply internal/gensupport/resumable.go]: I feel below condition is enough to check this `I think we'd only want to retry if we get deadline exceeded from rCtx, but ctx is still good.` Removed that condition.

File: internal/gensupport/retry.go
Comment (Tulsishah) [reply internal/gensupport/retry.go]: Done

File: googleapi/googleapi.go
Comment (tritone) [inline googleapi/googleapi.go:269]: Clarify: `// transfer timeout for resumable uploads.`

File: googleapi/googleapi.go
Comment (tritone) [inline googleapi/googleapi.go:271]: I would drop the sentence about existing clients; instead something like:

`// The default is no timeout for the request.`

File: internal/gensupport/resumable.go
Comment (tritone) [inline internal/gensupport/resumable.go:200]: I think since the default is zero, we don't need this if/else?

File: internal/gensupport/retry.go
Comment (tritone) [inline internal/gensupport/retry.go:30]: I think we don't need a variable for this given that zero is the language default for time.Duration.

File: internal/gensupport/resumable.go
Comment (tritone) [inline internal/gensupport/resumable.go:262]: I'm not sure this is a big deal practically speaking, but I'm not sure `defer cancel()` is what you want here, since that doesn't return until the function returns. So we might have a bunch of contexts stack up over the life of this loop.

Maybe we can explicitly cancel after transferChunk returns instead?

**Review by tritone — COMMENTED** [review]
Few more points here.

File: googleapi/googleapi.go
Comment (Tulsishah) [reply googleapi/googleapi.go]: Done

File: internal/gensupport/resumable.go
Comment (Tulsishah) [reply internal/gensupport/resumable.go]: I kept it this way to make it easier to update the default value from 0 to something else in the future.  Anyways updated the changes as you said.

File: internal/gensupport/retry.go
Comment (Tulsishah) [reply internal/gensupport/retry.go]: I kept it this way to make it easier to update the default value from 0 to something else in the future.  Anyways updated the changes as you said.

File: internal/gensupport/resumable.go
Comment (Tulsishah) [reply internal/gensupport/resumable.go]: Good catch. added the cancel when context is done.

File: googleapi/googleapi.go
Comment (tritone) [inline googleapi/googleapi.go:271]: You can remove the sentence about existing clients.

File: internal/gensupport/resumable.go
Comment (tritone) [inline internal/gensupport/resumable.go:268]: I think this would panic with chunkTransferTimeout = 0 because the function was never set to anything. Did you test this case?

File: internal/gensupport/resumable.go
Comment (tritone) [inline internal/gensupport/resumable.go:248]: Can you add a comment above saying what rCtx is for?

File: internal/gensupport/resumable.go
Comment (tritone) [inline internal/gensupport/resumable.go:266]: Use errors.Is()

**Review by tritone — COMMENTED** [review]
Few more comments here.

File: googleapi/googleapi.go
Comment (Tulsishah) [reply googleapi/googleapi.go]: Done

File: internal/gensupport/resumable.go
Comment (Tulsishah) [reply internal/gensupport/resumable.go:271]: Yes I have done manual testing, it is not having panic. I have assign rCtx=ctx and override the rCtx value in case of transfertimeout non zero. If it is zero then rCtx=ctx.

**Review by codyoss — COMMENTED** [review]
LGTM to me overall, I will wait for @tritone to fully sign-off. I will say this function is starting to get a bit long and complex. I would consider breaking up the loop body into some different functions can be better documented. Then you can use things like defer again too.

**Review by tritone — APPROVED** [review]

**Comment by tritone** [comment]
> LGTM to me overall, I will wait for @tritone to fully sign-off. I will say this function is starting to get a bit long and complex. I would consider breaking up the loop body into some different functions can be better documented. Then you can use things like defer again too.

It's a good point; I can do this in a subsequent PR.

**Review by codyoss — APPROVED** [review]

**Pull request merged**
